lld/lld-c/LLDAsLibraryC.cpp

```diff
@@ -732,68 +665,312 @@ resolveEVMLinkerSymbols(MemoryBufferRef inBuffer,
+  raw_string_ostream textSection(textSectionBuf);
+
+  // Define symbols whose values represent sizes of the dependencies.
+  // Skip the topLevelObject, as its size denotes .text section size of
+  // the resulting object file.
+  for (unsigned idx = 1; idx < memBufs.size(); ++idx) {
+    std::unique_ptr<Binary> binary =
+        cantFail(createBinary(unwrap(memBufs[idx])->getMemBufferRef()));
+    const auto *objFile = static_cast<const ObjectFile *>(binary.get());
+    std::string bufIdHash = EVM::getLinkerSymbolHash(depIDs[idx]);
+    if (depsToLink.count(bufIdHash))
+      getSymbolNameSections(objFile, depIDs[idx], symbolNameSectionsMap);
+
+    for (const SectionRef &sec : objFile->sections()) {
+      if (!sec.isText())
+        continue;
+
+      StringRef secName = cantFail(sec.getName());
+      if (secName == ".text") {
+        std::string sym = getDataSizeName(depIDs[idx]);
+        definedSymbols << sym << ";\n";
+        textSection << sym << " = " << sec.getSize() << ";\n";
+        break;
+      }
+    }
   }
 
-  if (!LLVMIsELFEVM(inBuffer))
-    return;
+  //   __dataoffset_Id_1 = .;
+  //   Id_1(.text);
+  //   __dataoffset_Id_2 = .;
+  //   Id_2(.text);
+  //   ...
+  //   __dataoffset_Id_N = .;
+  //   Id_N(.text);
+  for (unsigned idx = 0; idx < memBufs.size(); ++idx) {
+    std::string bufIdHash = EVM::getLinkerSymbolHash(depIDs[idx]);
+    // Do not link the dependency if it's not referenced via
+    // __dataoffset.
+    if (idx > 0 && !depsToLink.count(bufIdHash))
+      continue;
+
+    std::string sym = EVM::getDataOffsetSymbol(bufIdHash);
+    definedSymbols << sym << ";\n";
+    textSection << sym << " = ABSOLUTE(.);\n";
+    textSection << bufIdHash << "(.text);\n";
+  }
 
+  // Define a symbol whose value is the total size of the
+  // .text section.
+  std::string topDataSizeSym = getDataSizeName(depIDs[0]);
+  definedSymbols << topDataSizeSym << ";\n";
+  textSection << topDataSizeSym << " = ABSOLUTE(.);\n";
+
+  // It may be a case when we assemble several files that reference
+  // the same library which means they have the same sections
+  //
+  //   .symbol_name__linker_symbol__$[0-9a-f]{64}$__
+  //
+  //  By default, linker will concatenate them, which is not what we need.
+  //  We need to keep only one section with original content in the
+  //  output file. For this we drop all the duplicating input sections
+  //  and keep only one.
+  std::string discardSectionsBuf;
+  raw_string_ostream discardSections(discardSectionsBuf);
+  for (const auto &[secName, IDs] : symbolNameSectionsMap) {
+    // We need to discard all the symbol name sections, keeping
+    // only one of them (no matter which one exactly). Put
+    // these sections to the /DISCARD/ output section.
+    assert(IDs.size() > 0);
+    for (StringRef id : drop_begin(IDs)) {
+      std::string idHash = EVM::getLinkerSymbolHash(id);
+      discardSections << idHash << '(' << secName << ");\n";
+    }
+  }
+
+  std::string script =
```

Comment:
Could we improve readability here?
Raw sting literals _R"()"_ could help.

Reply:
Thanks, done.